sw/device/lib/base/freestanding/stdint.h

```diff
@@ -0,0 +1,141 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_BASE_FREESTANDING_STDINT_H_
+#define SW_DEVICE_LIB_BASE_FREESTANDING_STDINT_H_
+
+/**
+ * This header implements the stdint.h standard header, as required by C11 S4.6.
+ * This header is specified in detail in S7.20 the same.
+ *
+ * Actual limit values below are defined as:
+ * - Unsigned MAXs are defined as the |0xff..ffU| all-ones representation.
+ * - Signed MINs are defined as |0x80..00| all-zeroes-but-the-last
+ *   representation.
+ * - Unsigned MAXs are defined as |0x7f..ff| all-ones-but-the-last
+ *   representation.
+ * These shorthands are correct for two's complement and the ones least likely
+ * to get a mistake in.
+ *
+ * It goes without saying that this file assumes the underlying representation
+ * is two's complement and little-endian.
+ *
+ * See https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md
+ * for the RISC-V specifics this file conforms to.
+ */
+
+typedef unsigned char uint8_t;
+typedef unsigned short uint16_t;
+typedef unsigned int uint32_t;
+typedef unsigned long long uint64_t;
+typedef signed char int8_t;
+typedef signed short int16_t;
+typedef signed int int32_t;
+typedef signed long long int64_t;
+
+typedef uint8_t uint_least8_t;
+typedef uint16_t uint_least16_t;
+typedef uint32_t uint_least32_t;
+typedef uint64_t uint_least64_t;
+typedef int8_t int_least8_t;
+typedef int16_t int_least16_t;
+typedef int32_t int_least32_t;
+typedef int64_t int_least64_t;
+
+// NOTE: the "fast" integers should be at least the size of a machine word
+// (which is what most compilers define them as).
+typedef uint32_t uint_fast8_t;
+typedef uint32_t uint_fast16_t;
+typedef uint32_t uint_fast32_t;
+typedef uint64_t uint_fast64_t;
+typedef int32_t int_fast8_t;
+typedef int32_t int_fast16_t;
+typedef int32_t int_fast32_t;
+typedef int64_t int_fast64_t;
+
+typedef unsigned long uintptr_t;
+typedef signed long intptr_t;
+
+typedef uint64_t uintmax_t;
+typedef int64_t intmax_t;
+
+// clang-format off
+// The comment below is a sort of "nybble scale", which is intended to make
+// it easier to check that a particular literal is the length it should be
+// (especially the int64_t literals, which need to be exactly 16 digits long).
+//
+// Unfortunately, these must be literals rather than cute shift expressions,
+// since they must be usable in #if preprocessor directives.
+//
+// --------------- (0x0123456789abcdef)
+#define INT8_MIN   (0x80)
+#define INT16_MIN  (0x8000)
+#define INT32_MIN  (0x80000000L)
+#define INT64_MIN  (0x8000000000000000LL)
+#define INT8_MAX   (0x7f)
+#define INT16_MAX  (0x7fff)
+#define INT32_MAX  (0x7fffffffL)
+#define INT64_MAX  (0x7fffffffffffffffLL)
+#define UINT8_MAX  (0xffU)
+#define UINT16_MAX (0xffffU)
+#define UINT32_MAX (0xffffffffUL)
```

Comment:
Do we really need a `L` suffix here and in `INT32_MIN`?

Reply:
Yup. Good catch.